core/store/src/trie/mem/updating.rs

```diff
@@ -0,0 +1,1281 @@
+        self.updated_nodes.get_mut(index).unwrap().take().expect("Node taken twice")
+    }
+
+    /// Does the opposite of take_node; returns the node to the specified ID.
+    fn place_node(&mut self, index: UpdatedMemTrieNodeId, node: UpdatedMemTrieNode) {
+        assert!(self.updated_nodes[index].is_none(), "Node placed twice");
+        self.updated_nodes[index] = Some(node);
+    }
+
+    /// Creates a new updated node, assigning it a new ID.
+    fn new_updated_node(&mut self, node: UpdatedMemTrieNode) -> UpdatedMemTrieNodeId {
+        let index = self.updated_nodes.len();
+        self.updated_nodes.push(Some(node));
+        index
+    }
+
+    /// This is called when we need to mutate a subtree of the original trie.
+    /// It decrements the refcount of the original trie node (since logically
+    /// we are removing it), and creates a new node that is equivalent to the
+    /// original node. The ID of the new node is returned.
+    ///
+    /// If the original node is None, it is a marker for the root of an empty
+    /// trie.
+    fn convert_existing_to_updated(&mut self, node: Option<MemTrieNodeId>) -> UpdatedMemTrieNodeId {
+        match node {
+            None => self.new_updated_node(UpdatedMemTrieNode::Empty),
+            Some(node) => {
+                if let Some(trie_refcount_changes) = self.trie_refcount_changes.as_mut() {
+                    trie_refcount_changes.subtract(node.as_ptr(self.arena).view().node_hash(), 1);
+                }
+                self.new_updated_node(UpdatedMemTrieNode::from_existing_node_view(
+                    node.as_ptr(self.arena).view(),
+                ))
+            }
+        }
+    }
+
+    /// If the ID was old, converts it to an updated one.
+    fn ensure_updated(&mut self, node: OldOrUpdatedNodeId) -> UpdatedMemTrieNodeId {
+        match node {
+            OldOrUpdatedNodeId::Old(node_id) => self.convert_existing_to_updated(Some(node_id)),
+            OldOrUpdatedNodeId::Updated(node_id) => node_id,
+        }
+    }
+
+    fn add_refcount_to_value(&mut self, hash: CryptoHash, value: Option<Vec<u8>>) {
+        if let Some(trie_refcount_changes) = self.trie_refcount_changes.as_mut() {
+            trie_refcount_changes.add(hash, value.unwrap(), 1);
+        }
+    }
+
+    fn subtract_refcount_for_value(&mut self, hash: CryptoHash) {
+        if let Some(trie_refcount_changes) = self.trie_refcount_changes.as_mut() {
+            trie_refcount_changes.subtract(hash, 1);
+        }
+    }
+
+    /// Inserts the given key value pair into the trie.
+    pub fn insert(&mut self, key: &[u8], value: Vec<u8>) {
+        self.insert_impl(key, FlatStateValue::on_disk(&value), Some(value));
+    }
+
+    /// Inserts the given key value pair into the trie, but the value may be a reference.
+    /// This is used to update the in-memory trie only, without caring about on-disk changes.
+    pub fn insert_memtrie_only(&mut self, key: &[u8], value: FlatStateValue) {
```

Comment:
Can it be hidden by `#[test]`?

Reply:
Ahh, this is actually used for initial loading of flat state deltas :D because there we don't have the full key.